# Module quiz: The Building Blocks of a Program

1. What data type is the variable  x ?
   ```javascript
   var x = {};
   ```
   - Function
   - Array
   - Object
   ```
   Answer: Object
   Explanation: {} is the array literal.
   ```

2. What will be the output of running the following code?
   ```javascript
    try {
        console.log('hello)
    } catch(e) {
        console.log('caught')
    }
   ```
   - Uncaught SyntaxError: Invalid or unexpected token.
   - Caught
   ```
   Answer: Uncaught SyntaxError: Invalid or unexpected token.
   Explanation: You cannot handle SyntaxErrors in JavaScript using a try catch block.  
   ```

3. What value is printed when the following code runs?
   ```javascript
    var burger = ["bun", "beef", "lettuce", "tomato sauce", "onion", "bun"];
    console.log(burger[2]);
   ```
   - bun
   - beef
   - lettuce
   - tomato sauce
   - onion
   ```
   Answer: lettuce
   Explanation: Indices start at 0. Therefore lettuce  will print out as it is at index 2.
   ```

4. In the following code, how many methods does the `bicycle` object have?
   ```javascript
    var bicycle = {
        wheels: 2,
        start: function() {

        },
        stop: function() {

        }
    };
   ```
   - 1
   - 2
   - 3
   ```
   Answer: 2
   Explanation: wheels is a property and start and stop are methods.
   ```

5. When the following code runs, what will print out?
   ```javascript
    try {​
        throw new Error();​
        console.log('Hello');
    } catch(err) {​
        console.log('Goodbye');
    }
   ```
   - Hello
   - Goodbye
   ```
   Answer: Goodbye
   Explanation: When the error is thrown, the code will execute the code inside the catch statement.
   ```

6. If you mis-type some code in your JavaScript, what kind of error will that result in?
   - RangeError
   - SyntaxErrror
   - TypeError
   ```
   Answer: SyntaxErrror
   Explanation: 
   ```

7. Will the following code execute without an error?
   ```javascript
    function add(a, b) {
        console.log(a + b)​
    }​
    
    add(3, "4");
   ```
   - Yes
   - No
   ```
   Answer: Yes
   Explanation: Although the code has a bug, it will execute without error. Instead of adding the numbers and printing 7, it will print 34.
   ```

8. What will be printed when the following code runs?
   ```javascript
    var result;
    console.log(result);
   ```
   - undefined
   - null
   - 0
   ```
   Answer: undefined
   Explanation: Since no value was assigned to result, undefined will print out.
   ```

9. What will be the output of the following code?
   ```javascript
    var str = "Hello";
    str.match("jello");
   ```
   - null
   - undefined
   - empty string
   ```
   Answer: null
   Explanation: Since the string "Hello" doesn't contain the string "jello", it cannot be matched, and thus an array object with the matching result cannot be created. Hence, the return value is null, signaling the absence of a matching result.
   ```

10. What will be the output of the following code?
   ```javascript
    try {
        Number(5).toPrecision(300)
    } catch(e) {
        console.log("There was an error")
    }
   ```
   - RangeError
   - 5
   - e
   - "There was an error"
   ```
   Answer: "There was an error"
   Explanation: The catch block will execute and print because of the Range error.
   ```